stan/math/prim/fun/cholesky_factor_constrain.hpp

```diff
@@ -17,39 +17,39 @@ namespace math {
  * specified vector.  A total of (N choose 2) + N + (M - N) * N
  * elements are required to read an M by N Cholesky factor.
  *
- * @tparam T type of elements in the matrix
+ * @tparam T type of the vector
  * @param x Vector of unconstrained values
- * @param M Number of rows
- * @param N Number of columns
  * @return Cholesky factor
  */
-template <typename T>
-Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> cholesky_factor_constrain(
-    const Eigen::Matrix<T, Eigen::Dynamic, 1>& x, int M, int N) {
+template <typename T, require_eigen_vector_t<T>* = nullptr>
+Eigen::Matrix<value_type_t<T>, Eigen::Dynamic, Eigen::Dynamic>
+cholesky_factor_constrain(const T& x, int M, int N) {
   using std::exp;
+  using T_scalar = value_type_t<T>;
   check_greater_or_equal("cholesky_factor_constrain",
                          "num rows (must be greater or equal to num cols)", M,
                          N);
   check_size_match("cholesky_factor_constrain", "x.size()", x.size(),
                    "((N * (N + 1)) / 2 + (M - N) * N)",
                    ((N * (N + 1)) / 2 + (M - N) * N));
-  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> y(M, N);
-  T zero(0);
+  Eigen::Matrix<T_scalar, Eigen::Dynamic, Eigen::Dynamic> y(M, N);
+  T_scalar zero(0);
   int pos = 0;
 
+  const Eigen::Ref<const plain_type_t<T>>& x_ref = x;
   for (int m = 0; m < N; ++m) {
     for (int n = 0; n < m; ++n) {
-      y(m, n) = x(pos++);
+      y(m, n) = x_ref(pos++);
```

Comment:
Is there a way in Eigen to fill a lower triangular of a matrix with an array?

Reply:
I don't think so. I am changing this to use eigen expressions instead of inner loops.